hack/lib/golang.sh

```diff
@@ -283,32 +269,53 @@ EOF
   if [[ "${TRAVIS:-}" != "true" ]]; then
     local go_version
     go_version=($(go version))
-    if [[ "${go_version[2]}" < "go1.4" ]]; then
+    if [[ "${go_version[2]}" < "go1.6" ]]; then
       kube::log::usage_from_stdin <<EOF
 Detected go version: ${go_version[*]}.
-Kubernetes requires go version 1.4 or greater.
-Please install Go version 1.4 or later.
+Kubernetes requires go version 1.6 or greater.
+Please install Go version 1.6 or later.
 EOF
-      exit 2
+      return 2
     fi
   fi
+}
+
+# kube::golang::setup_env will check that the `go` commands is available in
+# ${PATH}. If not running on Travis, it will also check that the Go version is
+# good enough for the Kubernetes build.
+#
+# Inputs:
+#   KUBE_EXTRA_GOPATH - If set, this is included in created GOPATH
+#
+# Outputs:
+#   env-var GOPATH points to our local output dir
+#   env-var GOBIN is unset (we want binaries in a predictable place)
+#   env-var GO15VENDOREXPERIMENT=1
+#   current directory is within GOPATH
+kube::golang::setup_env() {
+  kube::golang::verify_go_version
+
+  kube::golang::create_gopath_tree
 
   GOPATH=${KUBE_GOPATH}
 
   # Append KUBE_EXTRA_GOPATH to the GOPATH if it is defined.
   if [[ -n ${KUBE_EXTRA_GOPATH:-} ]]; then
     GOPATH="${GOPATH}:${KUBE_EXTRA_GOPATH}"
   fi
-
-  # Append the tree maintained by `godep` to the GOPATH unless KUBE_NO_GODEPS
-  # is defined.
-  if [[ -z ${KUBE_NO_GODEPS:-} ]]; then
-    GOPATH="${GOPATH}:${KUBE_ROOT}/Godeps/_workspace"
-  fi
   export GOPATH
 
+  # Change directories so that we are within the GOPATH.  Some tools get really
+  # upset if this is not true.
+  local subdir
+  subdir=$(kube::realpath . | sed "s|$KUBE_ROOT||")
+  cd "${KUBE_GOPATH}/src/${KUBE_GO_PACKAGE}/${subdir}"
```

Comment:
The symlink _output stuff serves 2 purposes.

1) It allows you to 'make' or 'hack/build-go.sh' in any the kube dir, regardless of where it was untarred (good, IMO) - this means you do NOT have to be in the GOPATH yourself, but it means our scripts *do*.

2) It allows us to collect the final outputs of 'go install' in a known place rather than my own $GOPATH.

Once we have a fully dockerized build, (2) matters less.  I still think (1) is valuable (especially since I believe we will NOT get away from Makefile as our main build interface, but that is a diff topic).  So we *have* to change dirs like this.  We just need to make all our tools understand it.